controllers/space/space_controller.go

```diff
@@ -0,0 +1,343 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Reconciler reconciles a Space object
+type Reconciler struct {
+	Client         client.Client
+	MemberClusters map[string]cluster.Cluster
+}
+
+// SetupWithManager sets up the controller reconciler with the Manager and the given member clusters.
+// Watches the Space resources in the current (host) cluster as its primary resources.
+// Watches NSTemplateSets on the member clusters as its secondary resources.
+func SetupWithManager(mgr ctrl.Manager, memberClusters map[string]cluster.Cluster) error {
+	builder := ctrl.NewControllerManagedBy(mgr).
+		// watch Spaces in the host cluster
+		For(&toolchainv1alpha1.Space{}, builder.WithPredicates(predicate.GenerationChangedPredicate{}))
+	// watch NSTemplateSets in all the member clusters
+	for _, memberCluster := range memberClusters {
+		builder = builder.Watches(source.NewKindWithCache(&toolchainv1alpha1.NSTemplateSet{}, memberCluster.Cache),
+			&handler.EnqueueRequestForObject{},
+		)
+	}
+
+	return builder.Complete(&Reconciler{
+		Client:         mgr.GetClient(),
+		MemberClusters: memberClusters,
+	})
+}
+
+//+kubebuilder:rbac:groups=toolchain.dev.openshift.com,resources=spaces,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=toolchain.dev.openshift.com,resources=spaces/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=toolchain.dev.openshift.com,resources=spaces/finalizers,verbs=update
+
+// Reconcile ensures that there is an NSTemplateSet resource defined in the target member cluster
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("reconciling Space")
+
+	// Fetch the Space
+	space := &toolchainv1alpha1.Space{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, space)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errs.Wrap(err, "unable to get the current Space")
+	}
+	if !util.IsBeingDeleted(space) {
+		// Add the finalizer if it is not present
+		if err := r.addFinalizer(logger, space); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		requeue, err := r.ensureSpaceDeletion(logger, space)
+		return reconcile.Result{Requeue: requeue}, err
```

Comment:
we don't need to force to requeue immediately - if there was something deleted, then we should just wait for the event when the NSTemplateSet is finally gone.

Reply:
done in d34a855